grammar/generics.g

```diff
@@ -0,0 +1,35 @@
+Generics = "<" params:GenericParam* % "," ","? ">";
+GenericParam = attrs:OuterAttr* kind:GenericParamKind;
```

Comment:
That's more of a semantic restriction, to be really honest.
The introduction of `const` generics has made the old rules seem quite silly other thsn from a stylistic point of view. Most of the compiler has been fixed to not assume lifetimes come first.

Reply:
If it actually were a semantic restriction then this would not be an error:
```rust
macro_rules! foo {
    ($x: item) => {}
}

foo! {
    fn stuff<T, 'a>() {}
}
```

(Semantic vs. syntactic restrictions have semver consequences due to macros...)

And yeah... grammar wise this is super annoying, especially since const generic variables can be interspersed with type variables (afaik, or we haven't made any decision here...).

Reply:
I guess we currently have the check duplicated (since the AST type loses the ordering requirement). @petrochenkov knows more.

Reply:
I don't know anything anymore since @varkor rewrote everything for const generics.
Before that different kinds of parameters were stored separately in AST and the ordering was enforced solely in the parser.